zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -1,191 +1,272 @@
-                self.block_hash.to_be_bytes(),
-            ),
-            ("signal_root", None, self.signal_root.to_be_bytes()),
-            ("graffiti", None, self.graffiti.to_be_bytes()),
-            (
-                "prover+parentGasUsed+gasUsed",
-                None,
-                self.field9.to_be_bytes(),
-            ),
-            (
-                "blockMaxGasLimit+maxTransactionsPerBlock+maxBytesPerTxList",
-                None,
-                self.field10.to_be_bytes(),
-            ),
-        ]
+impl<F: Field> Default for PublicData<F> {
+    fn default() -> Self {
+        // has to have at least one history hash, block number must start with at least one
+        let mut ret = Self::new(&witness::Block::default());
+        ret.block_context.history_hashes = vec![U256::default()];
+        ret
     }
+}
 
-    /// get rpi bytes
-    pub fn rpi_bytes(&self) -> Vec<u8> {
-        self.assignments().iter().flat_map(|v| v.2).collect()
-    }
-
-    fn default<F: Default>() -> Self {
-        Self::new::<F>(&witness::Block::default())
-    }
-
-    /// create PublicData from block and taiko
-    pub fn new<F>(block: &witness::Block<F>) -> Self {
-        use witness::left_shift;
-        let field9 = left_shift(block.protocol_instance.prover, 96)
-            + left_shift(block.protocol_instance.parent_gas_used as u64, 64)
-            + left_shift(block.protocol_instance.gas_used as u64, 32);
-
-        let field10 = left_shift(block.protocol_instance.block_max_gas_limit, 192)
-            + left_shift(block.protocol_instance.max_transactions_per_block, 128)
-            + left_shift(block.protocol_instance.max_bytes_per_tx_list, 64);
-        PublicData {
-            l1_signal_service: block.protocol_instance.l1_signal_service.to_word(),
-            l2_signal_service: block.protocol_instance.l2_signal_service.to_word(),
-            l2_contract: block.protocol_instance.l2_contract.to_word(),
-            meta_hash: block.protocol_instance.meta_hash.hash().to_word(),
-            block_hash: block.protocol_instance.block_hash.to_word(),
-            parent_hash: block.protocol_instance.parent_hash.to_word(),
-            signal_root: block.protocol_instance.signal_root.to_word(),
-            graffiti: block.protocol_instance.graffiti.to_word(),
-            prover: block.protocol_instance.prover,
-            parent_gas_used: block.protocol_instance.parent_gas_used,
-            gas_used: block.protocol_instance.gas_used,
-            block_max_gas_limit: block.protocol_instance.block_max_gas_limit,
-            max_transactions_per_block: block.protocol_instance.max_transactions_per_block,
-            max_bytes_per_tx_list: block.protocol_instance.max_bytes_per_tx_list,
-            field9,
-            field10,
+impl<F: Field> PublicData<F> {
+    fn new(block: &witness::Block<F>) -> Self {
+        let meta_hash = Token::FixedBytes(block.protocol_instance.meta_hash.hash().to_word().to_be_bytes().to_vec());
```

Comment:
hash.hash() for what??

Reply:
From [protocol instance implementation](https://github.com/taikoxyz/zkevm-circuits/blob/5d785b444317048b29f3cdd9da3e2a599b1ab798/zkevm-circuits/src/witness/protocol_instance.rs#L51)

Reply:
meta_hash here is already a hash, no need hash again.

Reply:
in the old version meta_hash is a struct `MetaHash` which returns the actual hash when you call `hash()` on itself